packages/aws-cdk/lib/diff.ts

```diff
@@ -119,19 +119,16 @@ export function printSecurityDiff(
   oldTemplate: any,
   newTemplate: cxapi.CloudFormationStackArtifact,
   requireApproval: RequireApproval,
-  quiet?: boolean,
+  _quiet?: boolean,
   stackName?: string,
   changeSet?: DescribeChangeSetOutput,
   stream: FormatStream = process.stderr,
 ): boolean {
   const diff = fullDiff(oldTemplate, newTemplate.template, changeSet);
 
-  // must output the stack name if there are differences, even if quiet
-  if (!quiet || !diff.isEmpty) {
+  if (diffRequiresApproval(diff, requireApproval)) {
```

Comment:
Are you sure this isn't changing the behavior of `quiet`? We are bound to get another issue for that....

Reply:
What this was doing is always printing `Stack undefined` when `quiet` was not set, regardless of whether there was a diff to print. Followed up by the diff if the diff requires approval.

Right now, we only print the stack name if the diff requires approval, period, so it functions as a header.

I suppose the difference would be this:

| | quiet: false | quiet: true | Change |
|----|----|------------|----|
| Before (no diff) | stackName | (nothing) |
| After (no diff) | (nothing) | (nothing) | Now doesn't unnecesessarily print stackName anymore |
| Before (diff) | stackName, diff | diff |
| After (diff) | stackName, diff | stackName, diff | Now prints stackname where it didn't use to |

The new behavior seems more sensible, no?

Reply:
Closes https://github.com/aws/aws-cdk/issues/31685

Seems to be what https://github.com/aws/aws-cdk/pull/30186 tried to do but failed-ish.

Reply:
Proof:

```alloy
open util/boolean 

enum Change { PermBroadened, PermNarrowed }
enum ConfirmationSetting { Never, Broadened, Any }
enum Visible { StackName, Diff }

one sig Operation {
  quiet: Bool,
  changes: set Change,
  confirmationSetting: ConfirmationSetting,
  old: set Visible,
  new: set Visible,
}

pred diffRequiresApproval {
  Operation.confirmationSetting != Never
  (Operation.confirmationSetting = Broadened and PermBroadened in Operation.changes) or (Operation.confirmationSetting = Any and some Operation.changes)
}

pred isEmpty[o: Operation] { no o.changes }

fact {
  -- Old
  (Operation.quiet = False or not Operation.isEmpty) <=> StackName in Operation.old
  diffRequiresApproval <=> Diff in Operation.old

  -- New
  diffRequiresApproval <=> (StackName + Diff) = Operation.new
  not diffRequiresApproval <=> no Operation.new
}

check { 
 -- If the new code shows something, the old code also used to show it
 StackName in Operation.new => StackName in Operation.old
 Diff in Operation.new => Diff in Operation.old

 -- This is a restatement of what's above but just to be safe: 
  (Operation.confirmationSetting = Any and some Operation.changes) => Diff in Operation.new
  (Operation.confirmationSetting = Broadened and PermBroadened in Operation.changes) => Diff in Operation.new
}
```